File: src/utils/slugger.ts
/** A heading. This was taken from Astro's MarkdownHeading type */
export type Heading = {
  /** The depth of the heading element. A h1 would have a depth of 1, for example. */
  depth: number;
  /** The slug, or ID, of the heading element. */
  slug: string;
  /** The content of the heading element. */
  title: string;
};

// This class is based off of the BananaSlug class in github-slugger
export class Slugger {
  #occurrences: Map<string, number>;
  headings: Heading[];

  constructor() {
    this.#occurrences = new Map<string, number>();
    this.headings = [];
  }

  slug(heading: string, depth: number) {
    const cleanHeading = heading.replaceAll(/[^a-zA-Z0-9]/g, "-").toLowerCase();
    let slug = cleanHeading;
    if (this.#occurrences.has(cleanHeading)) {
      slug += `-${this.#occurrences.get(cleanHeading)}`;
      this.#occurrences.set(
        cleanHeading,
        this.#occurrences.get(cleanHeading)! + 1,
      );
    } else {
      this.#occurrences.set(cleanHeading, 1);
    }

    this.headings.push({ depth, slug, title: heading });
    return slug;
  }
}

File: src/utils/slugger.rs
use regex::Regex;
use serde::Serialize;
use std::collections::HashMap;

/// A heading. This was adapted from Astro's MarkdownHeading type
#[derive(Debug, Serialize)]
pub struct Heading {
    /// The depth of the heading element. A h1 would have a depth of 1, for example.
    pub depth: u8,
    /// The slug, or ID, of the heading element.
    pub slug: String,
    /// The content of the heading element.
    pub title: String,
}

pub struct Slugger {
    occurrences: HashMap<String, u64>,
    pub headings: Vec<Heading>,
}

impl Slugger {
    pub fn new() -> Slugger {
        Slugger {
            occurrences: HashMap::new(),
            headings: vec![],
        }
    }

    pub fn slug(&mut self, heading: String, depth: u8) -> String {
        let heading_re = match Regex::new(r"[^a-zA-Z0-9]") {
            Ok(re) => re,
            Err(_) => unreachable!(),
        };
        let clean_heading = heading_re.replace_all(&heading, "-").to_lowercase();
        let mut slug = clean_heading.clone();

        match self.occurrences.get(&clean_heading) {
            Some(occ) => {
                slug += &format!("-{occ}",);
                self.occurrences.insert(clean_heading.clone(), occ + 1);
            }
            None => {
                self.occurrences.insert(clean_heading.clone(), 1);
            }
        };

        self.headings.push(Heading {
            depth,
            slug: slug.clone(),
            title: heading,
        });

        return slug;
    }
}
